kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/expressions/InExpressionEvaluator.java

```diff
@@ -0,0 +1,344 @@
+      In in,
+      DataType valueDataType,
+      List<Expression> inListExpressions,
+      List<DataType> inListDataTypes) {
+    checkIsStringType(
+        valueDataType, in, "'IN' with collation expects STRING type for the value expression");
+    for (int i = 0; i < inListDataTypes.size(); i++) {
+      if (!isNullLiteral(inListExpressions.get(i))) {
+        checkIsStringType(
+            inListDataTypes.get(i),
+            in,
+            "'IN' with collation expects STRING type for all list elements");
+      }
+    }
+  }
+
+  private static void checkIsUTF8BinaryCollation(In in, CollationIdentifier collationIdentifier) {
+    if (!"SPARK.UTF8_BINARY".equals(collationIdentifier.toString())) {
+      throw unsupportedExpressionException(
+          in,
+          String.format(
+              "Unsupported collation: \"%s\". "
+                  + "Default Engine supports just \"SPARK.UTF8_BINARY\" collation.",
+              collationIdentifier));
+    }
+  }
+
+  private static void checkIsStringType(DataType dataType, In in, String message) {
+    if (!(dataType instanceof StringType)) {
+      throw unsupportedExpressionException(in, message);
+    }
+  }
+
+  @VisibleForTesting
+  static boolean areTypesCompatible(DataType type1, DataType type2) {
+    // Use the same type compatibility logic as equality comparisons
+    // to ensure semantic consistency: col IN (v1, v2) ≡ col = v1 OR col = v2
+    return type1.equivalent(type2) || canCastTo(type1, type2) || canCastTo(type2, type1);
+  }
+
+  /** Check if a DataType is a numeric type. */
+  private static boolean isNumericType(DataType type) {
+    return type instanceof ByteType
+        || type instanceof ShortType
+        || type instanceof IntegerType
+        || type instanceof LongType
+        || type instanceof FloatType
+        || type instanceof DoubleType
+        || type instanceof DecimalType;
+  }
+
+  @VisibleForTesting
+  static boolean compareValues(Object value1, Object value2, DataType valueType) {
+    if (value1 == null || value2 == null) {
+      return false;
+    }
+    // For numeric types, convert both values to the same type for comparison
+    if (isNumericType(valueType)) {
+      return compareNumericValues(value1, value2, valueType);
+    }
+    return getComparator(valueType).apply(value1, value2) == 0;
+  }
+
+  /**
+   * Compares numeric values by converting both to the target type. This method handles mixed
+   * numeric type comparisons by converting both values to the target type and using the appropriate
+   * comparator.
+   */
+  @VisibleForTesting
+  static boolean compareNumericValues(Object value1, Object value2, DataType targetTypeToCompare) {
+    Number num1 = (Number) value1;
+    Number num2 = (Number) value2;
+    Object convertedValue1 = convertToTargetTypeForCompare(num1, targetTypeToCompare);
+    Object convertedValue2 = convertToTargetTypeForCompare(num2, targetTypeToCompare);
```

Comment:
Doesn't this sometimes down-cast? Is that safe?

I think this was similar to my question before, should we either
(1) restrict all the set types to be the same or narrower than the valueType?
(2) up-cast all targetType to the widest type found in within the set + value?

Reply:
It seems like currently we
(1) make sure all the types are compatible (this can just mean all are numeric types)
and then
(2) cast all the set types to the valueType (may truncate!) and then compare